[Bug]: Flaky E2E: TestNICCounterIBDegradation fails when TestDCGMBootstrapCompletedAnnotation's DCGM pod deletion cordons the shared node

### Prerequisites

- [x] I searched existing issues
- [x] I can reproduce this issue

### Code of Conduct

- [x] I agree to follow NVSentinel's Code of Conduct

### Bug Description

## What happened

E2E job [`E2E Tests (AMD64 + PostgreSQL)`](https://github.com/NVIDIA/NVSentinel/actions/runs/28572910522/job/84714597007) failed on `main` (commit `f0a680dc`, an unrelated Dependabot docker bump):

```
--- FAIL: TestNICCounterIBDegradation/NIC_Counter_-_IB_Degradation (5.05s)
    nic-health-monitor.go:461: node nvsentinel-worker is still cordoned
    Error:      Condition satisfied
    Messages:   NIC cleanup should remain stable on node nvsentinel-worker
```

The failure is the `require.Never` stability check in `waitForNICCleanupComplete` (`tests/helpers/nic-health-monitor.go:454`): the node was clean when the test started, then got cordoned mid-way through the 10-second stability window.

## Root cause

The failing test is a result of test pollution from the **preceding** test, `TestDCGMBootstrapCompletedAnnotation` (added in #1425). Both tests pick the same worker node (`nvsentinel-worker`), and NVSentinel's fault pipeline is asynchronous, so a fault triggered by the first test lands as a cordon while the second test is running.

Timeline reconstructed from the job log and the `e2e-kind-container-logs` artifact (fault-quarantine / platform-connector logs):

| Time (UTC) | Event |
|---|---|
| 07:46:44.63 | `TestDCGMBootstrapCompletedAnnotation` deletes the `nvidia-dcgm` pod on `nvsentinel-worker` |
| 07:46:49.65 | Replacement DCGM pod Ready → test asserts annotation persisted, **passes and ends (no teardown)** |
| 07:46:49.68 | `TestNICCounterIBDegradation` starts on the same node; node is clean → `require.Eventually` passes; 10s `require.Never` stability window opens |
| 07:46:53.22 | gpu-health-monitor's periodic health check fires while its DCGM connection is still broken → emits fatal `GpuDcgmConnectivityFailure` (`DCGM_CONNECTIVITY_ERROR`, "Failed to connect to DCGM for health check") |
| 07:46:53.23 | fault-quarantine matches "GPU fatal error ruleset" → **cordons `nvsentinel-worker`** |
| 07:46:54.70 | `require.Never` tick sees `node.Spec.Unschedulable == true` → test fails |

Key fault-quarantine log line:

```json
{"time":"2026-07-02T07:46:53.228574212Z","level":"INFO","msg":"Cordoning node","node":"nvsentinel-worker"}
```
triggered by event `checkName=GpuDcgmConnectivityFailure, isFatal=true, errorCode=["DCGM_CONNECTIVITY_ERROR"], nodeName=nvsentinel-worker` (generated 07:46:53.21 — ~3.5s **after** the DCGM test already passed).

Notes:

- The post-bootstrap DCGM connectivity loss is intentionally fatal (that's the design from #1425). The bug is purely test isolation.
- `TestDCGMBootstrapCompletedAnnotation` only waits for the replacement pod to be Ready; it does not wait for the gpu-health-monitor's DCGM connection to recover, nor for the resulting quarantine to clear. Its sibling `TestGPUHealthMonitorDCGMConnectionError` (same file) *does* wait for the `GpuDcgmConnectivityFailure` condition to return to healthy and has a `Teardown` — the new test lacks both.
- The NIC test's defensive cleanup (`clearNICConditions`) only sends healthy **NIC** events, which cannot clear a **GPU**-caused quarantine, so it can't protect itself from this pollution.
- Flakiness window: it only reproduces when the gpu-health-monitor's periodic check timer fires between the DCGM pod deletion and connection recovery, and the resulting cordon lands inside the next test's 10s `Never` window. All E2E runs on `main` since #1425 merged (2026-06-29) passed until this one (2026-07-02).

## What I expected

E2E tests to be isolated: a passing test should not leave in-flight fatal health events that cordon the node used by the next test.

## Suggested fix

1. **Primary:** Add a recovery wait / `Teardown` to `TestDCGMBootstrapCompletedAnnotation` (`tests/gpu_health_monitor_test.go`): after the replacement DCGM pod is Ready, `require.Eventually` that the `GpuDcgmConnectivityFailure` condition is healthy and the node is not cordoned / has no active quarantine annotation — mirroring the existing pattern in `TestGPUHealthMonitorDCGMConnectionError`.
2. **Secondary hardening:** Make `waitForNICCleanupComplete` (or the NIC tests' setup) tolerate/clear pre-existing non-NIC quarantines before opening the stability window.


### Component

Other

### Steps to Reproduce

Deterministic reproduction (no need to wait for the flake): on a running E2E cluster, delete the `nvidia-dcgm` pod on a worker node and immediately run `TestNICCounterIBDegradation` targeting that node — the fatal `GpuDcgmConnectivityFailure` cordon lands during the stability window.

### Environment

- NVSentinel version: v1.12.0

### Logs/Output

Link to failed run: https://github.com/NVIDIA/NVSentinel/actions/runs/28572910522/job/84714597007

--- a/tests/gpu_health_monitor_test.go
+++ b/tests/gpu_health_monitor_test.go
@@ -23,6 +23,7 @@ import (
 	"fmt"
 	"strings"
 	"testing"
+	"time"
 
 	"tests/helpers"
 
@@ -43,6 +44,14 @@ const (
 	GPUHealthMonitorContainerName = "gpu-health-monitor"
 	GPUHealthMonitorDaemonSetName = "gpu-health-monitor-dcgm-4.x"
 )
+
+// dcgmConnectivityFailureObserveWindow bounds how long we wait for the fatal
+// GpuDcgmConnectivityFailure that the gpu-health-monitor emits on its first
+// poll after its DCGM connection breaks. The monitor polls every 15s
+// (PollIntervalSeconds in the gpu-health-monitor chart configmap), so three
+// poll cycles plus pipeline latency is ample; if nothing shows up by then the
+// connection survived and no fatal is in flight.
+const dcgmConnectivityFailureObserveWindow = 50 * time.Second
 
 const (
 	keyGpuHealthMonitorPodName      contextKey = "gpuHealthMonitorPodName"
@@ -957,6 +966,14 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 		testNodeName = gpuHealthMonitorPod.Spec.NodeName
 		t.Logf("Using node: %s", testNodeName)
 
+		// Deleting the nvidia-dcgm pod below breaks the gpu-health-monitor's DCGM
+		// connection; its next poll emits a fatal GpuDcgmConnectivityFailure, which
+		// would quarantine this node and leak a cordon into whichever test runs
+		// next. Mark the node unmanaged so fault-quarantine ignores that event.
+		t.Logf("Setting ManagedByNVSentinel=false on node %s", testNodeName)
+		err = helpers.SetNodeManagedByNVSentinel(ctx, client, testNodeName, false)
+		require.NoError(t, err, "failed to set ManagedByNVSentinel label")
+
 		node, err := helpers.GetNodeByName(ctx, client, testNodeName)
 		require.NoError(t, err, "failed to get node %s", testNodeName)
 		require.NotEmpty(t, node.Annotations[dcgmBootstrapAnnotation],
@@ -1000,6 +1017,75 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 
 			return ctx
 		})
+
+	feature.Teardown(func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+		if testNodeName == "" {
+			t.Log("Skipping teardown: node not selected (setup likely failed early)")
+			return ctx
+		}
+
+		client, err := c.NewClient()
+		require.NoError(t, err, "failed to create kubernetes client for teardown")
+
+		// The nvidia-dcgm pod restart leaves the gpu-health-monitor with a stale
+		// DCGM handle, so its next poll emits a fatal GpuDcgmConnectivityFailure.
+		// Before handing the node back as managed, wait for that failure to show
+		// up and for the monitor to reconnect and report healthy again — otherwise
+		// the fatal lands after this test ends and cordons the node under the
+		// next test (see the TestNICCounterIBDegradation flake).
+		t.Logf("Waiting up to %v for GpuDcgmConnectivityFailure to appear on node %s",
+			dcgmConnectivityFailureObserveWindow, testNodeName)
+		sawConnectivityFailure := false
+		deadline := time.Now().Add(dcgmConnectivityFailureObserveWindow)
+		for time.Now().Before(deadline) {
+			condition, err := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
+				"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsNotHealthy")
+			if err != nil {
+				t.Logf("Error checking condition: %v", err)
+			} else if condition != nil && condition.Status == v1.ConditionTrue {
+				sawConnectivityFailure = true
+				break
+			}
+			time.Sleep(helpers.WaitInterval)
+		}
+
+		if sawConnectivityFailure {
+			t.Logf("Waiting for GpuDcgmConnectivityFailure to become healthy on node %s", testNodeName)
+			require.Eventually(t, func() bool {
+				condition, err := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
+					"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsHealthy")
+				if err != nil {
+					t.Logf("Error checking condition: %v", err)
+					return false
+				}
+				return condition != nil && condition.Status == v1.ConditionFalse
+			}, helpers.EventuallyWaitTimeout, helpers.WaitInterval,
+				"GpuDcgmConnectivityFailure should become healthy on node %s", testNodeName)
+		} else {
+			t.Logf("No GpuDcgmConnectivityFailure observed within %v on node %s; "+
+				"gpu-health-monitor connection survived the nvidia-dcgm restart",
+				dcgmConnectivityFailureObserveWindow, testNodeName)
+		}
+
+		require.Never(t, func() bool {
+			condition, err := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
+				"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsNotHealthy")
+			if err != nil {
+				t.Logf("Error checking condition during stability check: %v", err)
+				return false
+			}
+			return condition != nil && condition.Status == v1.ConditionTrue
+		}, helpers.NeverWaitTimeout, helpers.WaitInterval,
+			"GpuDcgmConnectivityFailure should stay healthy on node %s", testNodeName)
+
+		t.Logf("Removing ManagedByNVSentinel label from node %s", testNodeName)
+		err = helpers.RemoveNodeManagedByNVSentinelLabel(ctx, client, testNodeName)
+		if err != nil {
+			t.Logf("Warning: failed to remove ManagedByNVSentinel label: %v", err)
+		}
+
+		return ctx
+	})
 
 	testEnv.Test(t, feature.Feature())
 }

--- a/tests/gpu_health_monitor_test.go
+++ b/tests/gpu_health_monitor_test.go
@@ -1030,9 +1030,7 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 		// The nvidia-dcgm pod restart leaves the gpu-health-monitor with a stale
 		// DCGM handle, so its next poll emits a fatal GpuDcgmConnectivityFailure.
 		// Before handing the node back as managed, wait for that failure to show
-		// up and for the monitor to reconnect and report healthy again — otherwise
-		// the fatal lands after this test ends and cordons the node under the
-		// next test (see the TestNICCounterIBDegradation flake).
+		// up and for the monitor to reconnect and report healthy again
 		t.Logf("Waiting up to %v for GpuDcgmConnectivityFailure to appear on node %s",
 			dcgmConnectivityFailureObserveWindow, testNodeName)
 		sawConnectivityFailure := false

--- a/tests/gpu_health_monitor_test.go
+++ b/tests/gpu_health_monitor_test.go
@@ -1071,32 +1071,16 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 
 		if sawConnectivityFailure {
 			t.Logf("Waiting for GpuDcgmConnectivityFailure to become healthy on node %s", testNodeName)
-			require.Eventually(t, func() bool {
-				condition, checkErr := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
-					"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsHealthy")
-				if checkErr != nil {
-					t.Logf("Error checking condition: %v", checkErr)
-					return false
-				}
-				return condition != nil && condition.Status == v1.ConditionFalse
-			}, helpers.EventuallyWaitTimeout, helpers.WaitInterval,
-				"GpuDcgmConnectivityFailure should become healthy on node %s", testNodeName)
+			helpers.WaitForNodeConditionWithCheckName(ctx, t, client, testNodeName,
+				"GpuDcgmConnectivityFailure", "", "GpuDcgmConnectivityFailureIsHealthy", v1.ConditionFalse)
 		} else {
 			t.Logf("No GpuDcgmConnectivityFailure observed within %v on node %s; "+
 				"gpu-health-monitor connection survived the nvidia-dcgm restart",
 				dcgmConnectivityFailureObserveWindow, testNodeName)
 		}
 
-		require.Never(t, func() bool {
-			condition, checkErr := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
-				"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsNotHealthy")
-			if checkErr != nil {
-				t.Logf("Error checking condition during stability check: %v", checkErr)
-				return false
-			}
-			return condition != nil && condition.Status == v1.ConditionTrue
-		}, helpers.NeverWaitTimeout, helpers.WaitInterval,
-			"GpuDcgmConnectivityFailure should stay healthy on node %s", testNodeName)
+		t.Logf("Verifying GpuDcgmConnectivityFailure stays healthy on node %s", testNodeName)
+		helpers.EnsureNodeConditionNotPresent(ctx, t, client, testNodeName, "GpuDcgmConnectivityFailure")
 
 		// Restore the original ManagedByNVSentinel label state.
 		if hadManagedLabel {

--- a/tests/gpu_health_monitor_test.go
+++ b/tests/gpu_health_monitor_test.go
@@ -954,6 +954,8 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 
 	var testNodeName string
 	var dcgmPodName string
+	var originalManagedLabel string
+	var hadManagedLabel bool
 
 	feature.Setup(func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
 		client, err := c.NewClient()
@@ -966,16 +968,20 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 		testNodeName = gpuHealthMonitorPod.Spec.NodeName
 		t.Logf("Using node: %s", testNodeName)
 
+		// Capture original label state so teardown can restore it exactly.
+		node, err := helpers.GetNodeByName(ctx, client, testNodeName)
+		require.NoError(t, err, "failed to get node %s", testNodeName)
+		originalManagedLabel, hadManagedLabel = node.Labels["k8saas.nvidia.com/ManagedByNVSentinel"]
+
 		// Deleting the nvidia-dcgm pod below breaks the gpu-health-monitor's DCGM
 		// connection; its next poll emits a fatal GpuDcgmConnectivityFailure, which
 		// would quarantine this node and leak a cordon into whichever test runs
 		// next. Mark the node unmanaged so fault-quarantine ignores that event.
-		t.Logf("Setting ManagedByNVSentinel=false on node %s", testNodeName)
+		t.Logf("Setting ManagedByNVSentinel=false on node %s (original: present=%v, value=%q)",
+			testNodeName, hadManagedLabel, originalManagedLabel)
 		err = helpers.SetNodeManagedByNVSentinel(ctx, client, testNodeName, false)
 		require.NoError(t, err, "failed to set ManagedByNVSentinel label")
 
-		node, err := helpers.GetNodeByName(ctx, client, testNodeName)
-		require.NoError(t, err, "failed to get node %s", testNodeName)
 		require.NotEmpty(t, node.Annotations[dcgmBootstrapAnnotation],
 			"node %s should have annotation %s set", testNodeName, dcgmBootstrapAnnotation)
 		t.Logf("Node %s has annotation %s: %s", testNodeName, dcgmBootstrapAnnotation,
@@ -1030,30 +1036,46 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 		// The nvidia-dcgm pod restart leaves the gpu-health-monitor with a stale
 		// DCGM handle, so its next poll emits a fatal GpuDcgmConnectivityFailure.
 		// Before handing the node back as managed, wait for that failure to show
-		// up and for the monitor to reconnect and report healthy again
+		// up and for the monitor to reconnect and report healthy again.
 		t.Logf("Waiting up to %v for GpuDcgmConnectivityFailure to appear on node %s",
 			dcgmConnectivityFailureObserveWindow, testNodeName)
 		sawConnectivityFailure := false
-		deadline := time.Now().Add(dcgmConnectivityFailureObserveWindow)
-		for time.Now().Before(deadline) {
-			condition, err := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
-				"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsNotHealthy")
-			if err != nil {
-				t.Logf("Error checking condition: %v", err)
-			} else if condition != nil && condition.Status == v1.ConditionTrue {
-				sawConnectivityFailure = true
-				break
+		hadSuccessfulCheck := false
+		ticker := time.NewTicker(helpers.WaitInterval)
+		defer ticker.Stop()
+		deadline := time.After(dcgmConnectivityFailureObserveWindow)
+	observeLoop:
+		for {
+			select {
+			case <-ctx.Done():
+				t.Logf("Context canceled while waiting for GpuDcgmConnectivityFailure on node %s", testNodeName)
+				break observeLoop
+			case <-deadline:
+				break observeLoop
+			case <-ticker.C:
+				condition, checkErr := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
+					"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsNotHealthy")
+				if checkErr != nil {
+					t.Logf("Error checking condition: %v", checkErr)
+					continue
+				}
+				hadSuccessfulCheck = true
+				if condition != nil && condition.Status == v1.ConditionTrue {
+					sawConnectivityFailure = true
+					break observeLoop
+				}
 			}
-			time.Sleep(helpers.WaitInterval)
 		}
+		require.True(t, hadSuccessfulCheck,
+			"never got a successful condition check for node %s during observe window", testNodeName)
 
 		if sawConnectivityFailure {
 			t.Logf("Waiting for GpuDcgmConnectivityFailure to become healthy on node %s", testNodeName)
 			require.Eventually(t, func() bool {
-				condition, err := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
+				condition, checkErr := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
 					"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsHealthy")
-				if err != nil {
-					t.Logf("Error checking condition: %v", err)
+				if checkErr != nil {
+					t.Logf("Error checking condition: %v", checkErr)
 					return false
 				}
 				return condition != nil && condition.Status == v1.ConditionFalse
@@ -1066,20 +1088,25 @@ func TestDCGMBootstrapCompletedAnnotation(t *testing.T) {
 		}
 
 		require.Never(t, func() bool {
-			condition, err := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
+			condition, checkErr := helpers.CheckNodeConditionExists(ctx, client, testNodeName,
 				"GpuDcgmConnectivityFailure", "GpuDcgmConnectivityFailureIsNotHealthy")
-			if err != nil {
-				t.Logf("Error checking condition during stability check: %v", err)
+			if checkErr != nil {
+				t.Logf("Error checking condition during stability check: %v", checkErr)
 				return false
 			}
 			return condition != nil && condition.Status == v1.ConditionTrue
 		}, helpers.NeverWaitTimeout, helpers.WaitInterval,
 			"GpuDcgmConnectivityFailure should stay healthy on node %s", testNodeName)
 
-		t.Logf("Removing ManagedByNVSentinel label from node %s", testNodeName)
-		err = helpers.RemoveNodeManagedByNVSentinelLabel(ctx, client, testNodeName)
-		if err != nil {
-			t.Logf("Warning: failed to remove ManagedByNVSentinel label: %v", err)
+		// Restore the original ManagedByNVSentinel label state.
+		if hadManagedLabel {
+			t.Logf("Restoring ManagedByNVSentinel=%s on node %s", originalManagedLabel, testNodeName)
+			require.NoError(t, helpers.SetNodeManagedByNVSentinel(ctx, client, testNodeName, originalManagedLabel == "true"),
+				"failed to restore ManagedByNVSentinel label on node %s", testNodeName)
+		} else {
+			t.Logf("Removing ManagedByNVSentinel label from node %s (was not present before test)", testNodeName)
+			require.NoError(t, helpers.RemoveNodeManagedByNVSentinelLabel(ctx, client, testNodeName),
+				"failed to remove ManagedByNVSentinel label from node %s", testNodeName)
 		}
 
 		return ctx

